### PR TITLE
Sync hotfix

### PR DESCRIPTION
**Pull request by ObsessiveOrange**

### CODE REVIEW

File: src/main/java/patching/PatchManager.java [hunk 1]
```diff
@@ -60,16 +76,24 @@ private void transformAndSendPatch(BatchingControl batchingCtrl, long fileID, IR
         // Create releaser to make sure it only is ever done once
         Runnable releaser = new Runnable() {
             final Object synchronizationObj = new Object();
-            boolean released = false;
+            volatile boolean released = false;
```
Comment (julie-is-late) [inline src/main/java/patching/PatchManager.java]: why does this need to be `volatile` if it's only ever accessed from within this runnable?

File: src/main/java/patching/PatchManager.java [hunk 2]
```diff
@@ -124,13 +153,16 @@ public void run() {
...
-                            logger.debug(String.format("PatchManager: Removing patches %s; patch done queue is currently %s", batchingCtrl.patchBatchingQueue.subList(0, patches.length), batchingCtrl.patchDoneQueue).replace("\n", "\\n") + "\n");
+                            logger.debug(String.format("PatchManager: Removing patches %s; patch queue is currently %s", batchingCtrl.patchBatchingQueue.subList(0, patches.length), batchingCtrl.patchBatchingQueue).replace("\n", "\\n"));
+                            logger.debug(String.format("PatchManager: Removing patches %s; patch done queue is currently %s", batchingCtrl.patchBatchingQueue.subList(0, patches.length), batchingCtrl.patchDoneQueue).replace("\n", "\\n"));
                             // Remove the sent patches
                             for (int i = 0; i < patches.length; i++) {
-                                batchingCtrl.patchDoneQueue.add(batchingCtrl.patchBatchingQueue.remove(0));
+                                Patch p = batchingCtrl.patchBatchingQueue.remove(0);
```
Comment (julie-is-late) [inline src/main/java/patching/PatchManager.java]: why are we catching this return in a variable if we're just going to throw it away?

Comment (julie-is-late) [reply src/main/java/patching/PatchManager.java]: Side note: now that we have `((FileChangeResponse) response.getData()).getChanges()`, do we still want to remove everything from the batching queue or just the accepted things? I don't think that we can't reject a single change currently, but I'm just wondering if it wouldn't be safer. 
Also, if we know it's going to return the full set, why even return them in the response to begin with?

File: src/main/java/patching/PatchManager.java [hunk 3]
```diff
@@ -148,57 +180,111 @@ public void run() {
...
-
-        if (!batchingByFile.containsKey(fileID)) {
-            BatchingControl batchingControl = new BatchingControl();
-            batchingByFile.put(fileID, batchingControl);
-            batchingControl.maxVersionSeen = -1;
-        }
-        BatchingControl batchingCtrl = batchingByFile.get(fileID);
+    // This has to be in a separate thread, so that we can have a queue, to make sure
+    // notifications are applied in the order they are received. Otherwise the waiting for
+    // no active change requests could come out in the wrong order.
```
Comment (julie-is-late) [inline src/main/java/patching/PatchManager.java]: fix the grammar. just drop the commas in the first sentence, but I'm not sure what the second sentence is even saying.

File: src/main/java/patching/PatchManager.java [hunk 4]
```diff
@@ -148,57 +180,111 @@ public void run() {
...
+    private void runNotificationHandlerThread() {
+        new Thread(() -> {
+            Thread.currentThread().setName("PatchManagerNotificationHandler");
+
+            while (true) {
+                Notification notification = null;
+                try {
+                    notification = notificationHandlerQueue.take();
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
```
Comment (julie-is-late) [inline src/main/java/patching/PatchManager.java]: we need to continue here or else we'll get a null pointer exception

File: src/main/java/patching/PatchManager.java [hunk 5]
```diff
@@ -148,57 +180,111 @@ public void run() {
...
-                    Patch donePatch = itr.next();
-                    if(donePatch.getBaseVersion() >= patch.getBaseVersion()){
-                        patch = patch.transform(donePatch);
-                    } else {
-                        // remove ones we don't care about anymore.
-                        itr.remove();
+                FileChangeNotification fileChangeNotif = (FileChangeNotification) notification.getData();
+                long fileID = notification.getResourceID();
+
+                BatchingControl batchingCtrl = getBatchingControl(fileID);
```
Comment (julie-is-late) [inline src/main/java/patching/PatchManager.java]: I know batching control is synchronized, but does it also need to be volatile if it is potentially being inserted into from this thread?

File: src/test/java/patching/TestPatchManager.java [hunk 6]
```diff
@@ -114,9 +116,83 @@ public void testNotificationHandler() throws IOException, ClassNotFoundException
         } catch (InterruptedException e) {
             e.printStackTrace();
         }
-
     }
 
+//    @Test
```
Comment (julie-is-late) [inline src/test/java/patching/TestPatchManager.java:122]: what's the plan with this test? why doesn't it work anymore, and is there something we need to be testing now instead?

File: src/main/java/patching/PatchManager.java [hunk 7]
```diff
@@ -148,57 +180,111 @@ public void run() {
...
+                if (batchingCtrl.activeChangeRequest) {
+                    synchronized (batchingCtrl) {
+                        while (batchingCtrl.activeChangeRequest) {
+                            try {
+                                batchingCtrl.wait();
+                            } catch (InterruptedException e) {
+                                e.printStackTrace();
```
Comment (julie-is-late) [inline src/main/java/patching/PatchManager.java]: we need to do some logic here so we don't get stuck if we catch this exception, currently it will just loop and probably keep throwing them

**Review by julie-is-late — CHANGES_REQUESTED** [review]

File: src/main/java/patching/PatchManager.java (hunk 2 above)
Comment (ObsessiveOrange) [reply src/main/java/patching/PatchManager.java]: There's actually no way for us to easily identify which patch was accepted and which failed. As of right now, the server accepts all or nothing, so this assumption is safe. We need to send it back so we can properly transform any old notifications.

File: src/main/java/patching/PatchManager.java (hunk 1 above)
Comment (ObsessiveOrange) [reply src/main/java/patching/PatchManager.java]: Done

File: src/main/java/patching/PatchManager.java (hunk 3 above)
Comment (ObsessiveOrange) [reply src/main/java/patching/PatchManager.java]: done

File: src/main/java/patching/PatchManager.java (hunk 4 above)
Comment (ObsessiveOrange) [reply src/main/java/patching/PatchManager.java]: done

File: src/main/java/patching/PatchManager.java (hunk 5 above)
Comment (ObsessiveOrange) [reply src/main/java/patching/PatchManager.java]: synchronization handles that. it forces a sync of the cache as well

File: src/main/java/patching/PatchManager.java (hunk 7 above)
Comment (ObsessiveOrange) [reply src/main/java/patching/PatchManager.java]: done. this is a wonky case, that has no "correct" solution, i don't think, but the best we can do is just apply the patch anyways.

File: src/test/java/patching/TestPatchManager.java (hunk 6 above)
Comment (ObsessiveOrange) [reply src/test/java/patching/TestPatchManager.java:122]: It's a test-case in progress, that we will need in future. I was trying to figure out how to test what i was looking for, and this didn't quite hit the edge case.

**Review by julie-is-late — DISMISSED** [review]

**Comment by julie-is-late** [comment]
@ObsessiveOrange this didn't pass travis

File: src/test/java/patching/TestDiff.java [hunk 8]
```diff
@@ -246,15 +246,26 @@ public void testTransform2A() {
...
+        diff1 = new Diff("4:+4:str1");
+        diff2 = new Diff("4:+4:str2");
+        result = diff2.transform(false, Collections.singletonList(diff1));
+        Assert.assertEquals(1, result.size());
+        Assert.assertEquals("4:+4:str2", result.get(0).toString());
+
+        diff1 = new Diff("0:+15:longTestString1");
+        diff2 = new Diff("0:+4:str2");
+        result = diff2.transform(false, Collections.singletonList(diff1));
+        Assert.assertEquals("0:+4:str2", result.get(0).toString());
```
Comment (julie-is-late) [inline src/test/java/patching/TestDiff.java:267]: are these the only tests we have of a `false` for precedence?

File: src/main/java/patching/PatchManager.java [hunk 9]
```diff
@@ -2,203 +2,433 @@
...
+
+        // Add to batching pre-queue, and then transfer to main BatchingQueue
+        // This avoids deadlocking the UI, or running the transformAndSendPatch on UI threads.
+        synchronized (batchingCtrl.patchBatchingPreQueue) {
+            logger.debug(String.format("PatchManager: Adding %s to batching pre-queue; batching pre-queue currently %s", Arrays.toString(patches), batchingCtrl.patchBatchingPreQueue).replace("\n", "\\n") + "\n");
+            Collections.addAll(batchingCtrl.patchBatchingPreQueue, patches);
         }
 
-        transformAndSendPatch(batchingCtrl, fileID, respHandler, sendErrHandler);
+        // Run the transformAndSendPatch on a new thread, to make sure the UI thread doesn't get blocked.
```
Comment (julie-is-late) [inline src/main/java/patching/PatchManager.java]: no comma

File: src/main/java/patching/PatchManager.java [hunk 10]
```diff
@@ -2,203 +2,433 @@
...
         wsMgr.sendAuthenticatedRequest(req);
-        timer.schedule(new TimerTask() {
-            @Override
-            public void run() {
+
+        // Wait for up to PATCH_TIMEOUT_MILLIS for the response. After which, assume network failure and try again.
+        try {
+            long startTime = System.currentTimeMillis();
+            if (!requestInFlightSem.tryAcquire(PATCH_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS)) {
                 logger.debug("PatchManager: Request timed out, running releaser.");
```
Comment (julie-is-late) [inline src/main/java/patching/PatchManager.java:221]: this debug is no longer true

File: src/main/java/patching/PatchManager.java [hunk 11]
```diff
@@ -2,203 +2,433 @@
...
+                synchronized (notificationHandlerQueue) {
+                    try {
+                        // Unlock writeLock when notification queue is drained.
+                        if (notificationHandlerQueue.isEmpty()) {
+                            if (hasWriteLock) {
+                                handlingNotificationsLock.writeLock().unlock();
+                                hasWriteLock = false;
+                            }
+                            // Put this thread to sleep while waiting for more notifications.
+                            notificationHandlerQueue.wait();
```
Comment (julie-is-late) [inline src/main/java/patching/PatchManager.java:273]: why are you waiting here? it seems weird to just wait until we have "enough" notifications to process any of them

**Review by julie-is-late — APPROVED** [review]

File: src/main/java/patching/PatchManager.java (hunk 10 above)
Comment (ObsessiveOrange) [reply src/main/java/patching/PatchManager.java:221]: It still is; If the tryAcquire fails to get a permit within the timeout, it returns false. It then logs that debug statement

File: src/main/java/patching/PatchManager.java (hunk 11 above)
Comment (ObsessiveOrange) [reply src/main/java/patching/PatchManager.java:273]: It's a wait/notify cycle. It waits so that the synchronization control is unlocked, and is notified by the notificationHandler when items are added into the queue

File: src/main/java/patching/PatchManager.java (hunk 9 above)
Comment (ObsessiveOrange) [reply src/main/java/patching/PatchManager.java]: changed

File: src/main/java/patching/PatchManager.java (hunk 10 above)
Comment (julie-is-late) [reply src/main/java/patching/PatchManager.java:221]: wait, where does it return false though?

**Comment by julie-is-late** [comment]
This looks good now 🎉🎉🎉
